gtsfm/frontend/cacher/matcher_cacher.py

```diff
@@ -0,0 +1,191 @@
+"""Cacher for any matcher, which caches the output on disk in the top level folder `cache`.
+
+This class provides the caching functionality to a GTSFM matcher. To use this cacher, initialize it with the matcher obj
+you want to apply the cache on.
+
+Example: To cache output of `TwoWayMatcher`, use `MatcherCacher(matcher_obj=TwoWayMatcher())`.
+
+Authors: Ayush Baid
+"""
+from pathlib import Path
+from typing import List, Optional, Tuple
+
+import numpy as np
+
+import gtsfm.utils.cache as cache_utils
+import gtsfm.utils.io as io_utils
+import gtsfm.utils.logger as logger_utils
+from gtsfm.common.keypoints import Keypoints
+from gtsfm.frontend.matcher.matcher_base import MatcherBase
+
+logger = logger_utils.get_logger()
+
+CACHE_ROOT_PATH = Path(__file__).resolve().parent.parent.parent.parent / "cache"
+NUM_KEYPOINTS_TO_SAMPLE_FOR_HASH = 10
+
+
+class MatcherCacher(MatcherBase):
+    """Cacher for matcher output on disk, keyed on the input."""
+
+    def __init__(self, matcher_obj: MatcherBase) -> None:
+        super().__init__()
+        self._matcher = matcher_obj
+        # TODO: make the obj cache key dependent on the code
+        self._matcher_obj_key = type(self._matcher).__name__
+
+    def __get_cache_path(self, cache_key: str) -> Path:
+        """Gets the cache path from the cache key."""
+        return CACHE_ROOT_PATH / "matcher" / "{}.pbz2".format(cache_key)
+
+    def __generate_cache_key(
+        self,
+        keypoints_i1: Keypoints,
+        keypoints_i2: Keypoints,
+        descriptors_i1: np.ndarray,
+        descriptors_i2: np.ndarray,
+        im_shape_i1: Tuple[int, int],
+        im_shape_i2: Tuple[int, int],
+    ) -> str:
+        """Generates the cache key from the input detections, image shapes, and underlying matcher.
+
+        This function uses the first few keypoints and descriptors to generate a key, and hence is not guaranteed to
+        be unique. However, since its unlikely that keypoints/descriptors exactly match in the selected few indices,
+        and differ in the remaining, its a good choice for a key.
+        """
+        numpy_arrays_to_hash: List[np.ndarray] = []
+
+        # subsample and concatenate keypoints and descriptors
+        numpy_arrays_to_hash.append(keypoints_i1.coordinates[:NUM_KEYPOINTS_TO_SAMPLE_FOR_HASH].flatten())
```

Comment:
@akshay-krishnan do you think it is simpler to just pass the full images to the matcher, and use SHA1 of the images? I wonder if we're making the hash key a bit too complicated below

Reply:
We do not have images as input, right?

Reply:
I think cleaner API (and hence cleaner plate notation and diagram) is our main priority. Even though this function is slightly messy, we probably wont change it for a long time.

Reply:
@ayushbaid  one potential method would be to do something like store the image hash key as an optional attribute of the `Keypoints` array. That way it would already be freely available here. Not saying we have to do it that way, but it's an option